core/src/main/java/com/linecorp/armeria/common/CompositeHttpHeadersBase.java

```diff
@@ -0,0 +1,224 @@
+
+        for (StringMultimapGetters<CharSequence, AsciiString> delegate : delegates()) {
+            final HttpHeaderGetters getter = (HttpHeaderGetters) delegate;
+            if (getter.isContentLengthUnknown()) {
+                return -1;
+            }
+
+            final long length = getter.contentLength();
+            if (length < 0) {
+                continue;
+            }
+
+            if (contentLength < 0) {
+                contentLength = 0;
+            }
+            contentLength += length;
+        }
+
+        return contentLength;
+    }
+
+    final void contentLengthUnknown() {
+        ((HttpHeadersBase) appender()).contentLengthUnknown();
+    }
+
+    @Override
+    public boolean isContentLengthUnknown() {
+        for (StringMultimapGetters<CharSequence, AsciiString> delegate : delegates()) {
+            if (((HttpHeaderGetters) delegate).isContentLengthUnknown()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    final void contentType(MediaType contentType) {
+        requireNonNull(contentType, "contentType");
+        ((HttpHeadersBase) appender()).contentType(contentType);
+    }
+
+    @Override
+    public @Nullable MediaType contentType() {
+        for (StringMultimapGetters<CharSequence, AsciiString> delegate : delegates()) {
+            final MediaType contentType = ((HttpHeaderGetters) delegate).contentType();
+            if (contentType != null) {
+                return contentType;
+            }
+        }
+        return null;
+    }
+
+    final void contentDisposition(ContentDisposition contentDisposition) {
+        requireNonNull(contentDisposition, "contentDisposition");
+        ((HttpHeadersBase) appender()).contentDisposition(contentDisposition);
+    }
+
+    @Override
+    public @Nullable ContentDisposition contentDisposition() {
+        for (StringMultimapGetters<CharSequence, AsciiString> delegate : delegates()) {
+            final ContentDisposition contentDisposition = ((HttpHeaderGetters) delegate).contentDisposition();
+            if (contentDisposition != null) {
+                return contentDisposition;
+            }
+        }
+        return null;
+    }
+
+    final void endOfStream(boolean endOfStream) {
+        ((HttpHeadersBase) appender()).endOfStream(endOfStream);
+    }
+
+    @Override
+    public boolean isEndOfStream() {
+        for (StringMultimapGetters<CharSequence, AsciiString> delegate : delegates()) {
+            if (((HttpHeaderGetters) delegate).isEndOfStream()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public final int hashCode() {
+        final int hashCode = super.hashCode();
+        return isEndOfStream() ? ~hashCode : hashCode;
+    }
+
+    @Override
+    public boolean equals(@Nullable Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (!(o instanceof HttpHeaderGetters)) {
```

Comment:
Is this intentional not using `CompositeHttpHeadersBase`?

Reply:
```java
// CompositeStingMultimap
        if (that instanceof CompositeStringMultimap ||
            that instanceof StringMultimap) { // allow StringMultimap to compare too
            for (NAME name : names()) {
                if (!getAll((IN_NAME) name.toString())
                        .equals(that.getAll((IN_NAME) name.toString()))) {
                    return false;
                }
            }
```
aha right. on `CompositeStingMultimap#equal`, I think we may can compare `CompositeStingMultimap` with `StringMultimap` so i intended to just check `!(o instanceof HttpHeaderGetters)` here, not `!(o instance of CompositeHttpHeadersBase )`.

Hmm should we allow only `o instanceof CompositeHttpHeadersBase ` to check equality here?
I'm not sure so please share your opinion~! thanks 🙇

Reply:
I got it! I added related test case :) (compare equality with CompositeHeader and just Header)